File: app/src/main/java/me/zhanghai/android/douya/network/AuthenticationInterceptor.java
/*
 * Copyright (c) 2017 Zhang Hai <Dreaming.in.Code.ZH@Gmail.com>
 * All Rights Reserved.
 */

package me.zhanghai.android.douya.network;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import java.io.IOException;
import java.net.HttpRetryException;

import okhttp3.Interceptor;
import okhttp3.Request;
import okhttp3.Response;
import okhttp3.ResponseBody;
import okhttp3.internal.http.UnrepeatableRequestBody;

// We cannot use Authenticator because it only retries on HTTP 401 (see
// RetryAndFollowUpInterceptor.java).
public abstract class AuthenticationInterceptor implements Interceptor {

    private int mMaxNumRetries;

    public AuthenticationInterceptor(int maxNumRetries) {
        mMaxNumRetries = maxNumRetries;
    }

    @Override
    public Response intercept(@NonNull Chain chain) throws IOException {

        Authenticator authenticator = getAuthenticator();
        Request request = authenticator.authenticate(chain.request());
        Response response = null;

        for (int retryCount = 0; retryCount < mMaxNumRetries; ++retryCount) {

            response = chain.proceed(request);
            if (response.isSuccessful()) {
                return response;
            }

            Request retryRequest = authenticator.retryAuthentication(response);
            if (retryRequest == null) {
                return response;
            }

            ResponseBody responseBody = response.body();
            if (responseBody != null) {
                responseBody.close();
            }
            if (retryRequest.body() instanceof UnrepeatableRequestBody) {
                throw new HttpRetryException("Cannot retry streamed HTTP body", response.code());
            }
            request = retryRequest;
        }

        return response;
    }

    protected abstract Authenticator getAuthenticator();

    public interface Authenticator {

        @NonNull
        Request authenticate(Request request) throws IOException;

        @Nullable
        Request retryAuthentication(Response response) throws IOException;
    }
}

File: app/src/main/java/me/zhanghai/android/douya/network/AuthenticationInterceptor.kt
/*
 * Copyright (c) 2019 Hai Zhang <dreaming.in.code.zh@gmail.com>
 * All Rights Reserved.
 */

package me.zhanghai.android.douya.network

import okhttp3.Interceptor
import okhttp3.Request
import okhttp3.Response
import java.io.IOException
import java.net.HttpRetryException

// We cannot use [okhttp3.Authenticator] because it only retries on HTTP 401.
// @see okhttp3.internal.http.RetryAndFollowUpInterceptor
abstract class AuthenticationInterceptor(
    private val maxRetries: Int
) : Interceptor {
    protected abstract var authenticator: Authenticator

    @Throws(IOException::class)
    override fun intercept(chain: Interceptor.Chain): Response {
        var retryCount = 0
        val authenticator = authenticator
        var request: Request = authenticator.authenticate(chain.request())
        while (true) {
            val response = chain.proceed(request)
            if (response.isSuccessful || retryCount >= maxRetries) {
                return response
            }
            request = authenticator.retryAuthentication(response) ?: return response
            response.body?.close()
            if (request.body?.isOneShot() == true) {
                throw HttpRetryException("Cannot retry with one-shot request body", response.code)
            }
            ++retryCount
        }
    }

    interface Authenticator {
        @Throws(IOException::class)
        fun authenticate(request: Request): Request

        @Throws(IOException::class)
        fun retryAuthentication(response: Response): Request?
    }
}
